### ISSUE TITLE
How can I serialize the auto-generated model classes myself?

### ISSUE THREAD

**Issue by hellosagar** [aws-amplify/amplify-android#815]
I try to send **auto-generated model class**   from one activity to another activity using intent through Gson as currently there is no way to parse this auto-generated model class as `Serializable` or as `Parcelable` so I have to convert it into String first using Gson
```kotlin
val intent = Intent(this,TestResultAddOrEditActivity::class.java)
intent.putExtra("testResult", Gson().toJson(item))
startActivity(intent)
```

which looks like this
![image](https://user-images.githubusercontent.com/50016799/92685339-24007180-f355-11ea-8ff6-cd8c8892a530.png)

but in another activity what I get is the null value of `createDate` and `date` which is an object of `Temporal.DateTime`
only this object is causing issue and I get other variables without any problem

Screenshot of object TestResult before serializing
![image](https://user-images.githubusercontent.com/50016799/92686023-7aba7b00-f356-11ea-9627-593140c93b69.png)
screenshot of TestResult object from the second activity after second after deserializing this serialized object using this code snippet
```kotlin
val intent = intent
testResult = Gson().fromJson(intent.getStringExtra("testResult"), TestResult::class.java)
```
and that's what the object of TestResult I get from this above code snippet
![image](https://user-images.githubusercontent.com/50016799/92686440-485d4d80-f357-11ea-9e07-2869a10e9efa.png)
This is quite surprising why I get `createDate` and `date` as null
I'll try yo give my best to explain this bug
Please tell me what's the problem
I'm waiting eagerly to get the solution to this problem
Thanks

**Comment by KayIlory** [comment]
I agree, this is a major BUG that got introduced in the version >1.3. @jamesonwilliams  any ideas would help

**Comment by hellosagar** [comment]
I think the problem is in the toString() method of Temporal.java line 266
![image](https://user-images.githubusercontent.com/50016799/92692358-a8f18800-f361-11ea-823a-a737e12d734f.png)

**Comment by richardmcclellan** [comment]
Hi @hellosagar and @Kayllory, a couple thoughts here:

- I think what's happening is that a `Temporal` object was constructed from a String that was an invalid format.  An improvement that I will try to make is to throw an exception instead of creating the `Temporal` object, so that you get a more helpful stacktrace that shows you what that invalid String is.  
- Currently though, the Temporal object is in fact created with a null `offsetDateTime`, which results in a `NullPointerException` if `toString()` is called.    In your case, `toString()` is called is because you are running the debugger, which calls the method in order to display the object - it may not be an issue during normal execution of your app.
 - If you'd like to serialize the object using gson, you'll need to make sure that your Gson() object has custom type adapters to (de)serialize the date types.  Internally, we use gson as well.  [Here is our serializer](https://github.com/aws-amplify/amplify-android/blob/28ab6f309e82a54d27056530e0b46216ab6678fa/aws-api-appsync/src/main/java/com/amplifyframework/api/aws/GsonVariablesSerializer.java#L45-L51), and [here is our deserializer](https://github.com/aws-amplify/amplify-android/blob/5abb4ca70b437e536c608655708a412a1403cc55/aws-api/src/main/java/com/amplifyframework/api/aws/GsonFactory.java#L34-L55), for reference.  It looks like these are currently public, though they were not really intended for customer use, so I'd suggest using the same logic to implement them in your own code.

So to summarize next steps for us:
 - `Temporal` object constructors that take a String should throw an exception on invalid input instead of creating an object with null `offsetDateTime`
 - Consider better ways to support serialization of the model types.

**Comment by KayIlory** [comment]
You are simply the best @richardmcclellan . We will work on the 3rd option of using how you serialize and deserialize

**Comment by hellosagar** [comment]
@richardmcclellan  with custom adapters deserialization and serialization is working fine.

**Comment by Scarlett13** [comment]
Hi @KayIlory @hellosagar could you share your deserializer class and how to use it? I could serialize it using serializer example from @richardmcclellan , but I couldnt implements the deserializer yet. Thanks.

**Comment by hellosagar** [comment]
This is the Deserializer class
`public class GsonDeserializer {

    private final Gson gson;

    public GsonDeserializer() {
        gson = new GsonBuilder()
                .registerTypeAdapter(Temporal.Date.class, new TemporalDateDeserializer())
                .registerTypeAdapter(Temporal.Time.class, new TemporalTimeDeserializer())
                .registerTypeAdapter(Temporal.Timestamp.class, new TemporalTimestampDeserializer())
                .registerTypeAdapter(Temporal.DateTime.class, new TemporalDateTimeDeserializer())
                .registerTypeAdapter(String.class, new StringDeserializer()).create();
    }

    public <T> Object deSerialize(String json, Class<T> object) {
        return gson.fromJson(json, object);
    }

    @Override
    public boolean equals(Object thatObject) {
        if (this == thatObject) {
            return true;
        }
        if (thatObject == null || getClass() != thatObject.getClass()) {
            return false;
        }
        return true;
    }

    @Override
    public int hashCode() {
        return 0;
    }

    private static final class TemporalDateDeserializer implements JsonDeserializer<Temporal.Date> {
        @Override
        public Temporal.Date deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
                throws JsonParseException {
            return new Temporal.Date(json.getAsString());
        }
    }

    private static final class TemporalDateTimeDeserializer implements JsonDeserializer<Temporal.DateTime> {
        @Override
        public Temporal.DateTime deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
                throws JsonParseException {
            return new Temporal.DateTime(json.getAsString());
        }
    }

    private static final class TemporalTimeDeserializer implements JsonDeserializer<Temporal.Time> {
        @Override
        public Temporal.Time deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
                throws JsonParseException {
            return new Temporal.Time(json.getAsString());
        }
    }

    private static final class TemporalTimestampDeserializer implements JsonDeserializer<Temporal.Timestamp> {
        @Override
        public Temporal.Timestamp deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
                throws JsonParseException {
            return new Temporal.Timestamp(json.getAsLong(), TimeUnit.SECONDS);
        }
    }

    private static final class StringDeserializer implements JsonDeserializer<String> {
        @Override
        public String deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context
        ) throws JsonParseException {
            if (json.isJsonPrimitive()) {
                return json.getAsJsonPrimitive().getAsString();
            } else if (json.isJsonObject()) {
                return json.toString();
            } else {
                throw new JsonParseException("Failed to parse String from " + json);
            }
        }
    }
}`

Use the `deSerialize` method to Deserialize
credit  to @KayIlory  and @richardmcclellan also

**Comment by KayIlory** [comment]
@richardmcclellan  @hellosagar  should we add a serializer/deserializer utility function? or that is an overkill?

**Comment by Scarlett13** [comment]
I tried to serialize my object using [this](https://github.com/aws-amplify/amplify-android/blob/28ab6f309e82a54d27056530e0b46216ab6678fa/aws-api-appsync/src/main/java/com/amplifyframework/api/aws/GsonVariablesSerializer.java#L45-L51) and call it in my activity like:

```
Map<String, Object> foo = new HashMap<>();
foo("Foo", this.foo);
String fooJson = new GsonVariablesSerializer().serialize(foo);
```
and it gave me string like so:
`{"Foo":{"createdAt":"2020-09-04T03:49:55.406Z","item":0 ...}`

and call deSerialize function like:
```
Foo foo = new GsonDeserializer().deSerialize(fooJson, Foo.class);
```

but when I try to de-serialize it using your code block, it give me null object.
`Foo:{id: null,createdAt:null...}` 

could you please help me pointing where did I doing it wrong? @richardmcclellan @hellosagar @KayIlory 

Thanks.

**Comment by MFRamon** [comment]
I've came across this, the simplest solution I figured was to serialize my Object into JSON using a Gson object. What I needed to achieve was to pass an object to another activity. This was the solution I did : 

Activity from which the object is being sent : 

```
Gson mGson = new Gson();
intent.putExtra("jsonObjectKey", mGson.toJson(objectInstance));
startActivity(intent);
```

Activity in which I'll receive that object : 

```
ObjectClass objectInstance;

Intent i = getIntent();
String objectStringJson = i.getStringExtra("event");
mGson = new Gson();
objectInstance = mGson.fromJson(objectStringJson, ObjectClass.class);
```

Thanks
@richardmcclellan @jamesonwilliams

**Comment by raphkim** [comment]
We are no longer prioritizing this issue in favor of Richard's workaround using Gson custom adapters for de/serializing auto-generated models. We will mark it to be auto-resolved.

**Comment by richardmcclellan** [comment]
To serialize a generated model to a String:

```
Todo todo = Todo.builder().name("Get started with Amplify").build();
String str = GsonFactory.instance().toJson(todo);
```

To deserialize it back to the original typed object:
```
Todo todo = GsonFactory.instance().fromJson(str, Todo.class);
```
